node/core/pvf/tests/it/main.rs

```diff
@@ -184,3 +190,52 @@ async fn execute_queue_doesnt_stall_if_workers_died() {
 		max_duration.as_millis()
 	);
 }
+
+#[tokio::test]
+async fn execute_queue_doesnt_stall_with_varying_executor_params() {
+	let host = TestHost::new_with_config(|cfg| {
+		cfg.execute_workers_max_num = 2;
+	});
+
+	let executor_params_1 = ExecutorParams::from(
+		&[ExecutorParam::Environment(ExecutionEnvironment::WasmtimeGeneric)][..],
+	);
+	let executor_params_2 = ExecutorParams::from(
+		&[
+			ExecutorParam::Environment(ExecutionEnvironment::WasmtimeGeneric),
+			ExecutorParam::StackLogicalMax(1024),
+		][..],
+	);
+
+	// Here we spawn 6 validation jobs for the `halt` PVF and share those between 2 workers. Every
+	// 3rd job will have different set of executor parameters. All the workers should be killed
+	// and in this case the queue should respawn new workers with needed executor environment
+	// without waiting. The jobs will be executed in 3 batches, each running two jobs in parallel,
+	// and execution time would be roughly 3 * TEST_EXECUTION_TIMEOUT
+	let start = std::time::Instant::now();
+	futures::future::join_all((0u8..=6).map(|i| {
+		host.validate_candidate(
+			halt::wasm_binary_unwrap(),
+			ValidationParams {
+				block_data: BlockData(Vec::new()),
+				parent_head: Default::default(),
+				relay_parent_number: 1,
+				relay_parent_storage_root: Default::default(),
+			},
+			match i % 3 {
+				0 => executor_params_1.clone(),
+				_ => executor_params_2.clone(),
+			},
+		)
+	}))
+	.await;
+
+	let duration = std::time::Instant::now().duration_since(start);
+	let max_duration = 3 * TEST_EXECUTION_TIMEOUT;
+	assert!(
+		duration >= max_duration,
```

Comment:
nit: Nice test! I would just add an upper bound as well, to make sure the duration actually is roughly a factor of 3. Right now it could be a factor of e.g. 6 and the test would pass.

Reply:
I thought about that myself, but I'm afraid of rendering the test flaky, especially in a CI environment. I can take a rather large spread like from `max_duration` and `2 * max_duration`, but would it make any sense then?

Reply:
Good point, but I kind of feel like if it takes `2 * max_duration`, that is not good, as it's twice the expected time. That could be a sign of a problem with the queue that should be looked into. 

I would do between 3 and 4 `* TEST_EXECUTION_TIMEOUT`. That should more than account for flakiness and I would be surprised if it took longer than that. This is all based on my unscientific gut feeling though.

Reply:
Thank you, that not just improved the test but also uncovered an error in this test: it ran 7 iterations instead of 6, and the overall time was greater than expected. Fixed.

Reply:
Nice! I hadn't noticed that, but it just shows that it's good to have strong test cases!